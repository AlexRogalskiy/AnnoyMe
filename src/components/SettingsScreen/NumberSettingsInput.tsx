import React from "react";
import { Text, TextInput, View } from "react-native";
import styles from "./styles";

interface Props {
  label: string;
  value: number;

  validate: (value: number) => boolean;
  onChange: (value: number) => void;

  testID?: string;
}

/** Input for numbers with validation. */
export default ({ label, value, validate, onChange }: Props) => (
  <View style={styles.inputContainer}>
    <Text style={styles.inputTitle}>{label}</Text>
    <TextInput
      style={[
        styles.input,
        validate(Number(value)) ? styles.inputValid : styles.inputInvalid,
      ]}
      defaultValue={(value || "").toString()}
      onChangeText={newValue => onChange(Number(newValue))}
      keyboardType="number-pad"
      testID="number-settings-input-text-input"
    />
  </View>
);
