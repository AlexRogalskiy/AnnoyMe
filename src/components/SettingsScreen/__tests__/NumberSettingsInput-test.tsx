import React from "react";
import renderer, { act } from "react-test-renderer";
import NumberSettingsInput from "../NumberSettingsInput";

const label = "Frequency";
const gt10 = (value: number) => value > 10;
const onChangeIgnore = (_: number) => undefined;

it("NumberSettingsInput renders input with valid number", () => {
  const tree = renderer
    .create(
      <NumberSettingsInput
        label={label}
        value={12}
        validate={gt10}
        onChange={onChangeIgnore}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot("NumberSettingsInput-valid");
});

it("NumberSettingsInput renders input with invalid number", () => {
  const tree = renderer
    .create(
      <NumberSettingsInput
        label={label}
        value={1}
        validate={gt10}
        onChange={onChangeIgnore}
      />,
    )
    .toJSON();

  expect(tree).toMatchSnapshot("NumberSettingsInput-invalid");
});

it("NumberSettingsInput changes number", () => {
  let changableValue: number | undefined;
  const onChange = (newValue: number) => (changableValue = newValue);

  const component = renderer.create(
    <NumberSettingsInput
      label={label}
      value={1}
      validate={gt10}
      onChange={onChange}
    />,
  ).root;

  const textInput = component.findByProps({
    testID: "number-settings-input-text-input",
  });
  expect(textInput).toBeTruthy();

  act(() => {
    textInput.props.onChangeText("22");
  });
  expect(changableValue).toBe(22);
});
