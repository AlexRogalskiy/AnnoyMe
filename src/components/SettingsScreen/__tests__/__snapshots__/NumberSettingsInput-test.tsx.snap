// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberSettingsInput renders input with invalid number: NumberSettingsInput-invalid 1`] = `
<View
  style={
    Object {
      "paddingBottom": 10,
    }
  }
>
  <Text
    style={
      Object {
        "fontWeight": "200",
      }
    }
  >
    Frequency
  </Text>
  <TextInput
    allowFontScaling={true}
    defaultValue="1"
    keyboardType="number-pad"
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
        },
        Object {
          "borderBottomColor": "#f44336",
        },
      ]
    }
    testID="number-settings-input-text-input"
    underlineColorAndroid="transparent"
  />
</View>
`;

exports[`NumberSettingsInput renders input with valid number: NumberSettingsInput-valid 1`] = `
<View
  style={
    Object {
      "paddingBottom": 10,
    }
  }
>
  <Text
    style={
      Object {
        "fontWeight": "200",
      }
    }
  >
    Frequency
  </Text>
  <TextInput
    allowFontScaling={true}
    defaultValue="12"
    keyboardType="number-pad"
    onChangeText={[Function]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "borderBottomWidth": 1,
        },
        Object {
          "borderBottomColor": "#c8e6c9",
        },
      ]
    }
    testID="number-settings-input-text-input"
    underlineColorAndroid="transparent"
  />
</View>
`;
